### ISSUE TITLE
Bad element type in SizeOf with generic calli instruction

### ISSUE THREAD

**Issue by ThadHouse** [dotnet/coreclr#14524]
I am trying to write a cross platform library that interops with native code, and was planning on using raw IL and the `calli` instruction to call into my native functions, to allow me to at runtime select the native library to load. I have used this in the past, where I would create individual shim functions for every call, and this worked. However, I was thinking, and reading through the IL spec it seems like it should be allowed, to just build some generic template to handle this code for me. However, when attempting to use `calli` with generic I receive a `BadImageFormatException: Bad element type in SizeOf` exception when the JIT attempts to compile the code. It seems like this should work, which makes it seem like a bug in the JIT. Below is both the generic code that seems like it should work, and also the same code without generics that does work.

```
	.method public static !!T Invoke<T, T2>(native int func, !!T2 val) cil managed
	{
		ldarg.1
		ldarg.0
		calli unmanaged cdecl !!0(!!1)
		ret
	}
```
I am calling the above with T as long, and T2 as int.

```
	.method public static int64 Invoke(native int func, int32 val) cil managed
	{
		ldarg.1
		ldarg.0
		calli unmanaged cdecl int64(int32)
		ret
	}
```

**Comment by jkotas** [comment]
Unmanaged calli is not supported with generic arguments. It is the same underlying reasons why DllImport is not supported with generics.

**Comment by sakno** [comment]
I see that [this](https://github.com/dotnet/runtime/pull/103) pull request allows to use blittable types. Is it applicable to `calli` instruction as well?

**Comment by jkotas** [comment]
Yes, the same rules should apply to calli.

**Comment by sakno** [comment]
So it is expected in .NET Core 3.1, right?

**Comment by jkotas** [comment]
No, .NET Core 3.1 is "done" for a while. All features we have been working on for last several months are for .NET 5.
